Issue with GZIP IO unit tests in Debian CI

Hi all,

i've added [debian autopkgtests](https://ci.debian.net/) to the skbio packages. in doing so I've noticed some strange test failures.

The output is below. Reproducing this exactly might be a bit of a pain, but if you have docker I can give you a dockerfile that should help.

Cheers,
K

----

## Python 2:
```
======================================================================
FAIL: test_open_gzip (skbio.io.tests.test_util.TestWriteBufferedReader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/skbio/io/tests/test_util.py", line 337, in test_open_gzip
    self.gzip_contents[10:])
AssertionError: "example_file\x00\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\xe4\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\xe4\xe4\xd2\xa2b\x85\xfc<\x85\xdc\xfc\xa2T\x85\x92\x8c\xc4< 'U!'3/\x95\x0b\x00\x8a\x9d\xf4\x165\x00\x00\x00" != "\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\xe4\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\xe4\xe4\xd2\xa2b\x85\xfc<\x85\xdc\xfc\xa2T\x85\x92\x8c\xc4< 'U!'3/\x95\x0b\x00\x8a\x9d\xf4\x165\x00\x00\x00"

======================================================================
FAIL: test_open_gzip_encoding (skbio.io.tests.test_util.TestWriteBufferedReader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/skbio/io/tests/test_util.py", line 360, in test_open_gzip_encoding
    self.gzip_encoded_contents[10:])
AssertionError: 'big5_file\x00[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00' != '[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00'

======================================================================
FAIL: test_open_gzip (skbio.io.tests.test_util.TestWriteBytesIO)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/skbio/io/tests/test_util.py", line 487, in test_open_gzip
    self.gzip_contents[23:])
AssertionError: "\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\xe4\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\xe4\xe4\xd2\xa2b\x85\xfc<\x85\xdc\xfc\xa2T\x85\x92\x8c\xc4< 'U!'3/\x95\x0b\x00\x8a\x9d\xf4\x165\x00\x00\x00" != "\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\xe4\xe4\xd2\xa2b\x85\xfc<\x85\xdc\xfc\xa2T\x85\x92\x8c\xc4< 'U!'3/\x95\x0b\x00\x8a\x9d\xf4\x165\x00\x00\x00"

======================================================================
FAIL: test_open_gzip_encoding (skbio.io.tests.test_util.TestWriteBytesIO)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/skbio/io/tests/test_util.py", line 496, in test_open_gzip_encoding
    self.gzip_encoded_contents[20:])
AssertionError: '[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00' != '\xe1\x05\x00\x00\x00'

======================================================================
FAIL: test_open_gzip (skbio.io.tests.test_util.TestWriteFilepath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/skbio/io/tests/test_util.py", line 337, in test_open_gzip
    self.gzip_contents[10:])
AssertionError: "example_file\x00\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\xe4\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\xe4\xe4\xd2\xa2b\x85\xfc<\x85\xdc\xfc\xa2T\x85\x92\x8c\xc4< 'U!'3/\x95\x0b\x00\x8a\x9d\xf4\x165\x00\x00\x00" != "\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\xe4\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\xe4\xe4\xd2\xa2b\x85\xfc<\x85\xdc\xfc\xa2T\x85\x92\x8c\xc4< 'U!'3/\x95\x0b\x00\x8a\x9d\xf4\x165\x00\x00\x00"

======================================================================
FAIL: test_open_gzip_encoding (skbio.io.tests.test_util.TestWriteFilepath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/skbio/io/tests/test_util.py", line 360, in test_open_gzip_encoding
    self.gzip_encoded_contents[10:])
AssertionError: 'big5_file\x00[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00' != '[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00'

----------------------------------------------------------------------
Ran 1790 tests in 55.480s

FAILED (SKIP=1, failures=6)
```

## Python 3
(looks superficially the same to me, but here it is anyway)

```
======================================================================
FAIL: test_open_gzip (skbio.io.tests.test_util.TestWriteBufferedReader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/skbio/io/tests/test_util.py", line 337, in test_open_gzip
    self.gzip_contents[10:])
AssertionError: b"example_file\x00\x0b\xc9\xc8,V\x00\xa2\xe[163 chars]\x00" != b"\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\x[147 chars]\x00"

======================================================================
FAIL: test_open_gzip_encoding (skbio.io.tests.test_util.TestWriteBufferedReader)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/skbio/io/tests/test_util.py", line 360, in test_open_gzip_encoding
    self.gzip_encoded_contents[10:])
AssertionError: b'big5_file\x00[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00' != b'[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00'

======================================================================
FAIL: test_open_gzip (skbio.io.tests.test_util.TestWriteBytesIO)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/skbio/io/tests/test_util.py", line 487, in test_open_gzip
    self.gzip_contents[23:])
AssertionError: b"\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\xe4[145 chars]\x00" != b"\xfc\xbc\x92\xd4\xbc\x12.\xcf\x12\x85\xfc\x[102 chars]\x00"

======================================================================
FAIL: test_open_gzip_encoding (skbio.io.tests.test_util.TestWriteBytesIO)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/skbio/io/tests/test_util.py", line 496, in test_open_gzip_encoding
    self.gzip_encoded_contents[20:])
AssertionError: b'[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00' != b'\xe1\x05\x00\x00\x00'

======================================================================
FAIL: test_open_gzip (skbio.io.tests.test_util.TestWriteFilepath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/skbio/io/tests/test_util.py", line 337, in test_open_gzip
    self.gzip_contents[10:])
AssertionError: b"example_file\x00\x0b\xc9\xc8,V\x00\xa2\xe[163 chars]\x00" != b"\x0b\xc9\xc8,V\x00\xa2\xe2\xfc\xdcT\x85\x[147 chars]\x00"

======================================================================
FAIL: test_open_gzip_encoding (skbio.io.tests.test_util.TestWriteFilepath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/skbio/io/tests/test_util.py", line 360, in test_open_gzip_encoding
    self.gzip_encoded_contents[10:])
AssertionError: b'big5_file\x00[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00' != b'[\xee\xb8,\x8f\x0b\x00&\x05\xd3\xe1\x05\x00\x00\x00'

----------------------------------------------------------------------
Ran 1999 tests in 67.198s

FAILED (failures=6)
```

--- a/skbio/io/tests/test_util.py
+++ b/skbio/io/tests/test_util.py
@@ -320,6 +320,13 @@ class WritableBinarySourceTests(object):
             self.assertTrue(result.closed)
             self.check_closed(file, True)
 
+    def compare_gzip_file_contents(self, a, b):
+        # The first 10 bytes of a gzip header include a timestamp. The header
+        # can be followed by other "volatile" metadata, so only compare gzip
+        # footers (last 8 bytes) which contain a CRC-32 checksum and the length
+        # of the uncompressed data.
+        self.assertEqual(a[-8:], b[-8:])
+
     def test_open_binary(self):
         self.check_open_state_contents(self.binary_file, self.binary_contents,
                                        True, encoding='binary',
@@ -332,9 +339,8 @@ class WritableBinarySourceTests(object):
         self.check_open_state_contents(self.gzip_file, self.text_contents,
                                        False, compression='gzip')
 
-        # The first 10 bytes of a gzip header include a timestamp, so skip.
-        self.assertEqual(self.get_contents(self.gzip_file)[10:],
-                         self.gzip_contents[10:])
+        self.compare_gzip_file_contents(self.get_contents(self.gzip_file),
+                                        self.gzip_contents)
 
     def test_open_bz2(self):
         self.check_open_state_contents(self.bz2_file, self.text_contents,
@@ -355,9 +361,9 @@ class WritableBinarySourceTests(object):
                                        self.decoded_contents, False,
                                        compression='gzip', encoding='big5')
 
-        # The first 10 bytes of a gzip header include a timestamp, so skip.
-        self.assertEqual(self.get_contents(self.gzip_encoded_file)[10:],
-                         self.gzip_encoded_contents[10:])
+        self.compare_gzip_file_contents(
+            self.get_contents(self.gzip_encoded_file),
+            self.gzip_encoded_contents)
 
     def test_open_bz2_encoding(self):
         self.check_open_state_contents(self.bz2_encoded_file,
@@ -482,18 +488,17 @@ class TestWriteBytesIO(WritableBinarySourceTests, WritableSourceTest):
         self.check_open_state_contents(self.gzip_file, self.text_contents,
                                        False, compression='gzip')
 
-        # The first 10 bytes of a gzip header include a timestamp, so skip.
-        self.assertEqual(self.get_contents(self.gzip_file)[10:],
-                         self.gzip_contents[23:])
+        self.compare_gzip_file_contents(self.get_contents(self.gzip_file),
+                                        self.gzip_contents)
 
     def test_open_gzip_encoding(self):
         self.check_open_state_contents(self.gzip_encoded_file,
                                        self.decoded_contents, False,
                                        compression='gzip', encoding='big5')
 
-        # The first 10 bytes of a gzip header include a timestamp, so skip.
-        self.assertEqual(self.get_contents(self.gzip_encoded_file)[10:],
-                         self.gzip_encoded_contents[20:])
+        self.compare_gzip_file_contents(
+            self.get_contents(self.gzip_encoded_file),
+            self.gzip_encoded_contents)
 
 
 class TestReadBufferedReader(ReadableBinarySourceTests, ReadableSourceTest):

